text/projects/Krzysztof_Pióro/Lyndon_Factorization.tex

```diff
@@ -0,0 +1,146 @@
+Dokładniej będziemy mieli trzy przypadki:
+\begin{itemize}
+    \item $w[j] = w[k]$: dodanie $w[j]$ do $v_2$ nie narusza założenia, że $v_2$ jest \textit{prawie proste}. Także w tym przypadku po prostu zwiększamy zmienne $j$ oraz $k$.
+    \item $w[j] > w[k]$: słowo $v_2 + w[j]$ staje się proste. W tym przypadku zwiększamy $j$ i ustawiamy $k$ na początek słowa $v_2$.
+    \item $w[j] < w[k]$: słowo $v_2 + w[j]$ przestaje być \textit{prawie proste}. W tym przypadku dla naszego słowa $v_2 = u^t\bar{u}$ dzielimy $u^t$ na $t$ słów \textit{prostych},
+    dodajemy je do $v_1$ (czyli wynikowej faktoryzacji Lyndona), ustawiamy zmienne $i$ oraz $k$ na początek pozostałej części słowa $v_2$ ($\bar{u}$), a zmienną $j$ na jedną pozycję dalej.
+\end{itemize}
+
+
+
+
+
+\begin{algorithm}[H]
+    \caption{\textbf{Duval$(w)$}}
+    \begin{algorithmic}
+        \State $i := 1$
+        \State factorization := empty list
+        \While {$i \leq n$}
+            \State $j := i+1;$  $k := i$;
+            \While {$j \leq n$ \textbf{and} $w[k] \leq w[j]$}
+                \If{$w[k] < w[j]$}
+                    \State $k := i$
+                \Else
+                    \State $k := k + 1$;
+                \EndIf
+                \State $j := j + 1$;
+            \EndWhile
+           \While {$i \leq k$}
+                \State factorization append $w[i\ldots i+j-k-1]$
+                \State $i := i + j - k$;
+           \EndWhile
+        \EndWhile
+        \State \Return factorization
+    \end{algorithmic}
+\end{algorithm}
+
+\begin{proof}[Dowód poprawności]
+    Z powyższych przypadków możemy od razu wywnioskować, że algorytm Duval'a rozkłada słowo $w$ na 
+    $w = w_1w_2\ldots w_k$ takie, że wszystkie $w_i$ są słowami \textit{prostymi}.
+    Pozostało pokazać, że $w_1 \geq w_2 \geq \cdots \geq w_k$. 
+    W tym celu zastanówmy się co dzieje się w trakcie kroku z trzeciego przypadku. Niech $a := w[j]$. 
+    Zauważmy, że $\bar{u}av < u$ dla 
+    dowolnego słowa $v$. Zatem wszystkie prefiksy słowa $v_2v_3$, które otrzymamy po kroku z trzeciego przypadku,
+    będą ściśle mniejsze od słowa $u$, czyli od ostatniego słowa z faktoryzacji słowa $v_1$. 
+    Z tego wynika, że słowa, które później będą dodane do faktoryzacji zachowają warunek monotoniczności.
+\end{proof}
+
+\begin{proof}[Złożoność algorytmu]    
+    Pierwsza wewnętrzna pętla w każdym kroku zwiększa zmienną $j$. Zmienna ta
+    może być jednak zmniejszana w wyniku trzeciego przypadku. Zauważmy jednak że, 
+    to o ile cofniemy tą zmienną wynosi co najwyżej tyle, jak długie jest poprzednio dodane słowo Lyndona. 
+    Dodatkowo za każdym razem kiedy cofamy zmienną $j$, ostatnie słowo w faktoryzacji jest inne, zatem
+    możemy oszacować sumę cofnięć przez $O(n)$. Otrzymujemy więc, że ta pętla działa w czasie $O(n)$.
+    
+    Druga wewnętrzna pętla w algorytmie działa sumarycznie w czasie $O(n)$, bo wypisuje faktoryzację.
+
+    Ponadto zewnętrzna pętla algorytmu nie przekroczy $n$ iteracji, 
```

Comment:
Seems like a theorem should be added, because I was wondering for a moment whether it's $O(n^2)$ or $O(n)$ in total

Reply:
Not sure what is the problem here. Is it possible that you were looking at old version of pdf. I've already changed last part from "co kończy dowód złożoności" to "co pokazuje, że cały algorytm działa w czasie $O(n)$"?